R/lucid.R

```diff
@@ -259,6 +259,292 @@ lucidClient <- function(service, authInfo) {
+      query <- list()
+      if (!is.null(applicationId))
+        query$application=applicationId
+      if (!is.null(from))
+        query$from = from
+      if (!is.null(until))
+        query$until = until
+      if (!is.null(interval))
+        query$interval = interval
+      handleResponse(GET(service, authInfo, path, queryString(query)))
+    },
+
+    getBundle = function(bundleId){
+      path <- paste("/bundles/", bundleId, sep="")
+      handleResponse(GET(service, authInfo, path))
+    },
+
+    updateBundleStatus = function(bundleId, status) {
+      path <- paste("/bundles/", bundleId, "/status", sep="")
+      json <- list()
+      json$status = status
+      handleResponse(POST_JSON(service, authInfo, path, json))
+    },
+
+    createBundle = function(application, content_type, content_length, checksum) {
+      json <- list()
+      json$application = application
+      json$content_type = content_type
+      json$content_length = content_length
+      json$checksum = checksum
+      handleResponse(POST_JSON(service, authInfo, "/bundles", json))
+    },
+
+    listApplications = function(accountId, filters = list()) {
+      path <- "/applications/"
+      query <- paste(filterQuery(
+        c("account_id", names(filters)),
+        c(accountId, unname(filters))
+      ), collapse = "&")
+      listRequest(service, authInfo, path, query, "applications")
+    },
+
+    getApplication = function(applicationId) {
+      path <- paste("/applications/", applicationId, sep="")
+      application <- handleResponse(GET(service, authInfo, path))
+
+      output_id <- application$content_id
+      path <- paste("/content/", output_id, sep="")
+
+      applications_output <- handleResponse(GET(service, authInfo, path))
+      application$url <- applications_output$url
+      application
+    },
+
+    getApplicationMetrics = function(applicationId, series, metrics, from=NULL, until=NULL, interval=NULL) {
+      path <- paste("/applications/", applicationId, "/metrics/", series, "/", sep="")
+      query <- list()
+      m <- paste(lapply(metrics, function(x){paste("metric", urlEncode(x), sep="=")}), collapse = "&")
+      if (!is.null(from))
+        query$from = from
+      if (!is.null(until))
+        query$until = until
+      if (!is.null(interval))
+        query$interval = interval
+      handleResponse(GET(service, authInfo, path, paste(m, queryString(query), sep="&")))
+    },
+
+    getLogs = function(applicationId, entries = 50) {
+      path <- paste0("/applications/", applicationId, "/logs")
+      query <- paste0("count=", entries, "&tail=0")
+      handleResponse(GET(service, authInfo, path, query))
+    },
+
+    createApplication = function(name, title, template, accountId) {
+      json <- list()
+      json$name <- name
+      # the title field is only used on connect
+      # the template and account id fields are not user by cloud.
+      # TODO: do we derive output type.
+      json$output_type <- "shiny"
```

Comment:
I left this as a todo - but ultimately this is largely derived from `appMode` in the manifest. Which I think is derived here,

https://github.com/rstudio/rsconnect/blob/6801917d55d9d4035fbf643a6f4be3a6f795c764/R/bundle.R#L236

But wiring that through seems a bit messy - so I was wondering if there's a cleaner way to get that here that I'm missing?

Reply:
Hm. Connect gets its content type indirectly from the uploaded bundle. It's not provided when the content is created. Content starts off with an "unknown" type, which receives its type with the first bundle upload and is thereafter prevented from changing.

How is the "template" argument used (currently always given the value "shiny")? Connect doesn't have that notion.

Reply:
Yeah - we followed a path more similar to how connect handles this. Which removes this as a parameter. This is now determined at build time from the bundle rather then as the bundle is uploaded.